### PR TITLE
Add automatic pull secrets handling in kubernetes provider

### PR DESCRIPTION
**Pull request by pbialon**
Fix ENG-1843

### CODE REVIEW

File: pkg/kubernetes/run.go [hunk 1]
```diff
@@ -139,8 +140,11 @@ func (k *kubernetesDriver) runContainer(
 
 	// service account
 	serviceAccount := ""
-	if k.options.ServiceAccount != "" {
+	if k.options.ServiceAccount != "" || k.options.KubernetesPullSecretsEnabled == "true" {
```
Comment (FabianKramm) [inline pkg/kubernetes/run.go:143]: I don't think this is correct, as it will try to create the default service account here

Comment (pbialon) [reply pkg/kubernetes/run.go:143]: Hmmm, we need a service account for pull secrets (we need to use that SA for pods using this container registry). 

What would be your approach to ensure we have **some** service account here?

File: pkg/docker/config.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package docker
+
+import (
+	"os"
+	"path/filepath"
+)
+
+func ConfigPath() string {
+	dockerConfigPath := filepath.Join(os.Getenv("HOME"), ".docker", "config.json")
```
Comment (FabianKramm) [inline pkg/docker/config.go:9]: Please use the same functionality as DevPod uses:
```
package docker

import (
	"os"
	"path/filepath"

	"github.com/docker/cli/cli/config"
	"github.com/docker/cli/cli/config/configfile"

	"github.com/docker/docker/pkg/homedir"
)

const dockerFileFolder = ".docker"

func LoadDockerConfig() (*configfile.ConfigFile, error) {
	configDir := os.Getenv("DOCKER_CONFIG")
	if configDir == "" {
		configDir = filepath.Join(homedir.Get(), dockerFileFolder)
	}

	return config.Load(configDir)
}

```

File: pkg/kubernetes/pullsecrets.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+) error {
+	//todo: check if pull secret already exists, don't create it again
+	err := k.createPullSecret(ctx, *pullSecretMetaData, dockerConfigPath)
+	if err != nil {
+		return err
+	}
+
+	k.Log.Debugf("Pull secret '%s' created", pullSecretMetaData.pullSecretName)
+
+	err = k.addImagePullSecretsToServiceAccount(ctx, *pullSecretMetaData)
```
Comment (FabianKramm) [inline pkg/kubernetes/pullsecrets.go:22]: Don't add the pull secret to the service account, instead set it directly on the pod

File: pkg/kubernetes/pullsecrets.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	ctx context.Context,
+	psMetaData PullSecretMetaData,
+	dockerConfigPath string,
+) error {
+	args := []string{
+		"create",
+		"secret",
+		"generic",
+		psMetaData.pullSecretName,
+		"--from-file=.dockerconfigjson=" + dockerConfigPath,
```
Comment (FabianKramm) [inline pkg/kubernetes/pullsecrets.go:47]: that is not correct unfortunately, you need to extract a token (like devspace / devpod does) and then create a config out of that

File: pkg/kubernetes/pullsecrets.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+	out, err := k.buildCmd(ctx, args).CombinedOutput()
+	if err != nil {
+		return perrors.Wrapf(err, "create pull secret: %s", string(out))
+	}
+
+	return nil
+}
+
+func (k *kubernetesDriver) addImagePullSecretsToServiceAccount(
```
Comment (FabianKramm) [inline pkg/kubernetes/pullsecrets.go:59]: Again don't add to service account

**Review by FabianKramm — CHANGES_REQUESTED** [review]

File: pkg/kubernetes/run.go [hunk 6]
```diff
@@ -164,6 +168,29 @@ func (k *kubernetesDriver) runContainer(
...
+		}
+
+		dockerCredentials, err := docker.GetAuthConfig(host)
+		if err != nil {
+			return fmt.Errorf("load docker dockerCredentials: %w", err)
+		}
+
+		err = k.EnsurePullSecret(ctx, getPullSecretsName(id), dockerCredentials)
+		k.Log.Infof("Pull secret '%s' created", getPullSecretsName(id))
+
```
Comment (FabianKramm) [inline pkg/kubernetes/run.go:188]: Remove new line here

File: pkg/kubernetes/run.go [hunk 7]
```diff
@@ -141,6 +142,9 @@ func (k *kubernetesDriver) runContainer(
 	serviceAccount := ""
 	if k.options.ServiceAccount != "" {
 		serviceAccount = k.options.ServiceAccount
+		if serviceAccount == "" {
```
Comment (FabianKramm) [inline pkg/kubernetes/run.go:145]: Remove this as this is dead code

File: pkg/kubernetes/run.go [hunk 8]
```diff
@@ -164,6 +168,29 @@ func (k *kubernetesDriver) runContainer(
 	// parse resources
 	resources := parseResources(k.options.Resources, k.Log)
 
+	// ensure pull secrets
+	if k.options.KubernetesPullSecretsEnabled == "true" {
+		k.Log.Debugf("Ensure pull secrets")
+
+		//todo: host
```
Comment (FabianKramm) [inline pkg/kubernetes/run.go:175]: What does this mean?

File: pkg/kubernetes/pullsecrets.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/loft-sh/devpod-provider-kubernetes/pkg/docker"
+	perrors "github.com/pkg/errors"
+)
+
+func (k *kubernetesDriver) EnsurePullSecret(
+	ctx context.Context,
+	pullSecretName string,
+	dockerCredentials *docker.Credentials,
+) error {
+	//todo: check if pull secret already exists, don't create it again
```
Comment (FabianKramm) [inline pkg/kubernetes/pullsecrets.go:15]: How do we prevent this from causing issues right now?

File: pkg/kubernetes/run.go [hunk 10]
```diff
@@ -164,6 +168,29 @@ func (k *kubernetesDriver) runContainer(
...
+		if err != nil {
+			return fmt.Errorf("get registry from image name: %w", err)
+		}
+
+		dockerCredentials, err := docker.GetAuthConfig(host)
+		if err != nil {
+			return fmt.Errorf("load docker dockerCredentials: %w", err)
+		}
+
+		err = k.EnsurePullSecret(ctx, getPullSecretsName(id), dockerCredentials)
```
Comment (FabianKramm) [inline pkg/kubernetes/run.go:186]: Pull secret is created but never deleted

File: pkg/kubernetes/pullsecrets.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+) error {
+	args := []string{
+		"create",
+		"secret",
+		"docker-registry",
+		pullSecretName,
+		"--docker-username=" + dockerCredentials.Username,
+		"--docker-password=" + dockerCredentials.Secret,
+		"--docker-server=" + dockerCredentials.ServerURL,
+	}
```
Comment (FabianKramm) [inline pkg/kubernetes/pullsecrets.go:36]: I would actually prefer the DevSpace way here, as username etc might be empty often, see above in the docker config.go. If you chose to go this way, we definitely need an e2e test for the case no username and username and password for this before we can merge your PR

**Review by FabianKramm — CHANGES_REQUESTED** [review]

**Review by FabianKramm — APPROVED** [review]

**Pull request merged**